bluetooth: ots: Add object write procedure

ekohandel opened the pull request:
Add the ability to perform an OTS Write Procedure with only Patch
support configuration.

Signed-off-by: Abe Kohandel <abe.kohandel@gmail.com>

Thalley commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -255,6 +351,58 @@ static void oacp_read_proc_execute(struct bt_ots *ots,
+
+	rc = ots->cb->obj_write(ots, conn, ots->cur_obj->id, buf->data, len,
+				  offset, rem);
+	if (rc == 0)
+		if (offset + len > ots->cur_obj->metadata.size.cur)
+			ots->cur_obj->metadata.size.cur = offset + len;
```
Zephyr style guide states that even one-liner if statement shall use curly brackets

ekohandel replied:
I am surprised the style check passed :(! Sorry about that!

Thalley commented:
Very nice addition, just a few comments

ekohandel commented on subsys/bluetooth/services/ots/ots.c:
```diff
@@ -36,8 +36,23 @@ LOG_MODULE_REGISTER(bt_ots, CONFIG_BT_OTS_LOG_LEVEL);
 #define OACP_FEAT_BIT_READ 0
 #endif
 
+#if defined(CONFIG_BT_OTS_OACP_WRITE_SUPPORT)
+#define OACP_FEAT_BIT_WRITE BIT(BT_OTS_OACP_FEAT_WRITE)
+#else
+#define OACP_FEAT_BIT_WRITE 0
+#endif
+
+#if defined(CONFIG_BT_OTS_OACP_PATCH_SUPPORT)
+#define OACP_FEAT_BIT_PATCH BIT(BT_OTS_OACP_FEAT_PATCH)
+#else
+#define OACP_FEAT_BIT_PATCH 0
+#endif
```
I am not overwriting anything here, just using these defines to construct a bitmask to easily verify the features requested via the `bt_ots_init` function are a subset of the features enabled via the Kconfig configurations:
```
/* OACP features supported by Kconfig */
#define OACP_FEAT (		\
	OACP_FEAT_BIT_READ |	\
	OACP_FEAT_BIT_WRITE |	\
	OACP_FEAT_BIT_PATCH
```
and then later:
```
/* Check OACP supported features against Kconfig. */
if (ots_init->features.oacp & (~((uint32_t) OACP_FEAT))) {
	return -ENOTSUP;
}
```
I followed the same pattern that was being used for both `OACP_FEAT` and `OALP_FEAT`.

Thalley replied:
Oh, sorry, I had somehow read this as redefining the defines in ots.h :)

ekohandel commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -65,6 +86,74 @@ static enum bt_gatt_ots_oacp_res_code oacp_read_proc_validate(
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
+
+	/* append is not supported */
+	if ((params->offset + (uint64_t) params->len) > ots->cur_obj->metadata.size.alloc) {
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
```
This is supposed to cover the following error condition from table 3.18:
>The sum of the values of the Offset and Length parameters exceeds
the value of the Allocated Size field of the Object Size characteristic
AND the Server does NOT support appending additional data to an
object.

Also later after table 3.18 the specification says:
>If the Server supports appending additional data to an object, as indicated by the relevant bit of the OTS
Feature characteristic, and the sum of the values of the Offset and Length parameters exceeds the value
of the Allocated Size field of the Object Size characteristic, and none of the error conditions listed in Table
3.18 has occurred, the Server shall proceed as described in Section 3.3.2.6.1

I assumed the term "appending" in table 3.18 referred to the actual append capability. In the case that the append capability is supported, the application would be the one that would determine there aren't enough resources to actually append to the object and return the `BT_GATT_OTS_OACP_RES_INSUFF_RES` status.

Thalley replied:
Ah, yes, I missed that error condition from the table. This should be fine then :)

ekohandel commented on subsys/bluetooth/services/ots/ots_oacp.c:
```diff
@@ -65,6 +86,74 @@ static enum bt_gatt_ots_oacp_res_code oacp_read_proc_validate(
+		return BT_GATT_OTS_OACP_RES_OBJ_LOCKED;
+	}
+
+	ots->l2cap.rx_done = oacp_write_proc_cb;
+	ots->l2cap.closed = oacp_l2cap_closed;
+	ots->cur_obj->state.type = BT_GATT_OTS_OBJECT_WRITE_OP_STATE;
+	ots->cur_obj->state.write_op.recv_len = 0;
+	memcpy(&ots->cur_obj->state.write_op.oacp_params, params,
+		sizeof(ots->cur_obj->state.write_op.oacp_params));
```
Yeah, I followed the same pattern as read, I would like to keep it the same for now. We can improve this in a non functional change PR later.

I have more changes coming to support object creation so I am motivated to continue improving this implementation.

ekohandel commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -134,11 +134,30 @@ static uint32_t ots_obj_read(struct bt_ots *ots, struct bt_conn *conn,
+			     uint64_t id, uint8_t *data, uint32_t len,
+			     uint32_t offset, uint32_t rem)
+{
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	uint32_t obj_index = (id % ARRAY_SIZE(objects));
+
+	bt_ots_obj_id_to_str(id, id_str, sizeof(id_str));
```
Can I keep it the same for this PR and improve this as part of a non functional change PR?

Thalley replied:
Yes :)

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -548,6 +548,31 @@ struct bt_ots_cb {
+	 *  @param conn   The connection that wrote object.
+	 *  @param id     Object ID.
+	 *  @param data   Next chunk of data to be written.
+	 *  @param len    Length of the current chunk of data in the buffer
```
```suggestion
	 *  @param len    Length of the current chunk of data in the buffer.
```

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -548,6 +548,31 @@ struct bt_ots_cb {
+	 *  @param rem    Remaining length in the write operation.
+	 *
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  @return -EINPROGRESS has a special meaning and is unsupported at the
+	 *                       moment. It should not be returned.
```
I guess we could/should add a check for this value when calling this callback, and at least log a warning or error

Thalley commented on subsys/bluetooth/services/ots/ots.c:
```diff
@@ -184,6 +232,11 @@ int bt_ots_obj_add(struct bt_ots *ots,
 	}
 
+	CHECKIF(!ots_obj_validate_prop_against_oacp(obj_init->props, ots->features.oacp)) {
+		LOG_DBG("Object properties are not a subset of OACP");
```
Might be worth logging the values of obj_init->props and ots->features.oacp, so it will possible to debug.

Thalley reviewed (changes requested):
A few minor comments, otherwise LGTM

ekohandel commented:
Complete set of tests based on Table 3.18:

### The Write Op Code is not supported by the Server. (Op Code Not Supported)
Verify write is not supported by the server:
```
[Zephyr OTS]# gatt.select-attribute 00002abd-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0011]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011 Value:
  10 00 00 00 01 00 00 00                          ........
  10 00 00 00 01 00 00 00                          ........
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0011]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 02
```
Returned error code 0x02 (Op Code Not Supported)

### The Current Object is an Invalid Object. (Invalid Object)
Select OACP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 05
```
Returned error code 0x05 (Invalid Object)

### The object’s properties do not permit writing to the object. (Procedure Not Permitted)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000101 without write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x01 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 08                                         `..
```
Returned error code 0x08 (Procedure Not Permitted)

### Patching was attempted but patching is not supported by the Server. (Procedure Not Permitted)
Verify patching is not supported by the server:
```
[Zephyr OTS]# gatt.select-attribute 00002abd-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0011]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011 Value:
  30 00 00 00 01 00 00 00                          0.......
  30 00 00 00 01 00 00 00                          0.......
```
Select OCLP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0011]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  0c 00 00 00                                      ....
  0c 00 00 00                                      ....
```
Check the object's size:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac0-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0017]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017 Value:
  32 00 00 00 64 00 00 00                          2...d...
  32 00 00 00 64 00 00 00                          2...d...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0017]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 08                                         `..
```
Returned error code 0x08 (Procedure Not Permitted)

### Patching was attempted but the object’s properties do not permit patching of the object contents. (Procedure Not Permitted)
Verify patching is supported by the server:
```
[Zephyr OTS]# gatt.select-attribute 00002abd-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0011]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0011 Value:
  30 01 00 00 01 00 00 00                          0.......
  30 01 00 00 01 00 00 00                          0.......
```
Select OCLP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0011]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property but no patch property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property but no patch property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  0c 00 00 00                                      ....
  0c 00 00 00                                      ....
```
Check the object's size:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac0-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0017]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017 Value:
  32 00 00 00 64 00 00 00                          2...d...
  32 00 00 00 64 00 00 00                          2...d...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0017]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 08                                         `..
```
Returned error code 0x08 (Procedure Not Permitted)

### Truncation was attempted but the object’s properties do not permit truncation of the object contents. (Procedure Not Permitted)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0 with truncation mode bit set:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x02"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 08                                         `..
```
Returned error code 0x08 (Procedure Not Permitted)

### An Object Transfer Channel was not available for use. (Channel Unavailable)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Perform write procedure of 1 byte at offset 0
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 06                                         `..
```
Returned error code 0x06 (Channel Unavailable)

### The Mode parameter contains an RFU value. (Invalid Parameter)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Open a L2CAP Channel via a custom program:
```
sudo ./l2cap_sample -b D3:BF:16:24:E5:08 -w 001122
```
Perform write procedure of 1 byte at offset 0 with an RFU mode bit set:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x01"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 03                                         `..
Returned error code 0x03 (Invalid Parameter)
```

### The value of the Offset parameter exceeds the value of the Current Size field of the Object Size characteristic.  (Invalid Parameter)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Check the object's size:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac0-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0017]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017 Value:
  32 00 00 00 64 00 00 00                          2...d...
  32 00 00 00 64 00 00 00                          2...d...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0017]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Open a L2CAP Channel via a custom program:
```
sudo ./l2cap_sample -b D3:BF:16:24:E5:08 -w 001122
```
Perform write procedure of 1 byte at offset 0x33 which exceeds Current Size field:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x33 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 03                                         `..
Returned error code 0x03 (Invalid Parameter)
```

### The sum of the values of the Offset and Length parameters exceeds the value of the Allocated Size field of the Object Size characteristic AND the Server does NOT support appending additional data to an object.  (Invalid Parameter)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Check the object's size:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac0-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0017]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017 Value:
  32 00 00 00 64 00 00 00                          2...d...
  32 00 00 00 64 00 00 00                          2...d...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0017]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Open a L2CAP Channel via a custom program:
```
sudo ./l2cap_sample -b D3:BF:16:24:E5:08 -w 001122
```
Perform write procedure of 0x33 bytes at offset 0x32 which exceeds Allocated Size field:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x32 0x00 0x00 0x00 0x33 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 03                                         `..
Returned error code 0x03 (Invalid Parameter)
```

### The value of the Length parameter exceeds the number of octets that the Server has the capacity to write to the object.  (Insufficient Resources)
This is an application layer error and is not tested.

### The Current Object is locked by the Server. Object Locked An object transfer is already in progress that is using the Current Object. (Object Locked)
Select OCLP attribute and enable notifications:
```
[Zephyr OTS]# gatt.select-attribute 00002ac6-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0020]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Notifying: yes
Notify started
```
Select object 0x000000000100 with write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.write "0x05 0x00 0x01 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0020 Value:
  70 05 01                                         p..
```
Verify object has write property:
```
[Zephyr OTS:/service0010/char0020]# gatt.select-attribute 00002ac4-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001b]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001b Value:
  4c 00 00 00                                      L...
  4c 00 00 00                                      L...
```
Check the object's size:
```
[Zephyr OTS:/service0010/char001b]# gatt.select-attribute 00002ac0-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char0017]# gatt.read
Attempting to read /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char0017 Value:
  32 00 00 00 64 00 00 00                          2...d...
  32 00 00 00 64 00 00 00                          2...d...
```
Select OACP attribute and enable notifications:
```
[Zephyr OTS:/service0010/char0017]# gatt.select-attribute 00002ac5-0000-1000-8000-00805f9b34fb
[Zephyr OTS:/service0010/char001d]# gatt.notify on
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Notifying: yes
Notify started
```
Open a L2CAP Channel via a custom program:
```
sudo ./l2cap_sample -b D3:BF:16:24:E5:08 -w 001122
```
Perform write procedure of 0x10 bytes at offset 0x32:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x06 0x32 0x00 0x00 0x00 0x10 0x00 0x00 0x00 0x00"
Attempting to write /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 06 01                                         `..
```
Write operation is successful, data transfer is pending over the L2CAP socket, now attempt a write of 0x01 byte at offset 0x00 before the write operation is completed:
```
[Zephyr OTS:/service0010/char001d]# gatt.write "0x05 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00"
Attempting to write /org/bluez/hci1/dev_D3_BF_16_24_E5_08/service0010/char002d
[CHG] Attribute /org/bluez/hci0/dev_D3_BF_16_24_E5_08/service0010/char001d Value:
  60 05 09
```
Returned error code 0x09 (Object Locked)

### The requested procedure failed for a reason other than those enumerated above in this table.  (Operation Failed)
Unused error code so not tested.

joerchan commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -102,11 +117,20 @@ static void l2cap_connected(struct bt_l2cap_chan *chan)
 
 static const struct bt_l2cap_chan_ops l2cap_ops = {
 	.sent		= l2cap_sent,
+	.recv		= l2cap_recv,
```
Is alloc_buf callback not needed for OTS? Is OTS using L2CAP CoC segmentation, or defining its own segmentation?

There is a configuration for BT_OTS_L2CAP_CHAN_RX_MTU, but that is not allowed to exceed the size of the L2CAP PDU MTU. So if OTS wants to received L2CAP SDUs segmented into multiple L2CAP PDUs then the callback should be defined and allocate buffers from a net buf pool sized from BT_OTS_L2CAP_CHAN_RX_MTU.

ekohandel replied:
I have to admit I am not very familiar with this aspect of Bluetooth. From what I see not having the `alloc_buf` callback is forcing the MTU to be truncated.
```
<wrn> bt_l2cap: Segmentation disabled but MTU > MPS, truncating MTU
```
It seems more correct to add the `alloc_buf` callback to allow the MTU to be dictated by the OTS layer.

joerchan replied:
So you are sending and receiving L2CAP SDUs. The maximum size of the SDU is the MTU.
These packets are segmented into L2CAP PDUs. The maximum size of the PDU is the MPS.

By not having an alloc_buf callback then a single packet received in the l2cap_recv callback can me a maximum of MTU = MPS - 2.

By adding an alloc_buf callback then the application determines the maximum MTU by setting `chan->rx.mtu = CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU`.
The warning you saw there was the stack truncating the MTU to fit inside the MPS, so that option has not been  useful so far.

You could potentially skip having the alloc callback if the MPS is large enough for the OTS RX MTU.
`#if CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU <=  BT_L2CAP_SDU_RX_MTU`
 
The thing I couldn't quite work out by having a quick look at the OTS specification is whether OTS does segmentation on top of the L2CAP SDU again. But from what I'm seeing in this PR it looks like you can use offset and len to write an object that is larger than the SDU.
But from what I can tell there must be segmentation on top of the L2CAP SDU because the SDU length field is 16-bit, and the objects length field is 32-bit.

Thalley replied:
Yes, for large objects (i.e. objects larger than the L2CAP MTU), OTS supports an additional level of fragmentation.

ekohandel replied:
I think for simplicity I will just leave the allocation callback.

ekohandel replied:
I have added conditional usage of the allocation callback only if `CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU > BT_L2CAP_SDU_RX_MTU`.

joerchan commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -90,6 +90,21 @@ static void l2cap_sent(struct bt_l2cap_chan *chan)
 	}
 }
 
+int l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
```
Just a heads up about: https://github.com/zephyrproject-rtos/zephyr/pull/35958

Let me know if there is something that you would like to see changed in that PR.

ekohandel replied:
I think it looks fine as is, depending on which PR goes in first, there may be a conflict to resolve but I think without the actual receive functionality built into OTS the change you have makes sense.

joerchan commented on subsys/bluetooth/services/ots/ots_oacp_internal.h:
```diff
@@ -80,7 +88,7 @@ struct bt_gatt_ots_oacp_proc {
 			uint32_t len;
 			uint8_t mode;
-		} write_params;
+		} __packed write_params;
```
If they are all protocol definitions then they should probably all have __packed then?

ekohandel replied:
On a 32-bit processor the outcome is no different but you are correct, it should be packed to ensure correct behavior on all architecture types.

jhedberg replied:
Since we don't have endianness-specific integer types like Linux, `__packed` also acts as a reminder that the integer struct members are in some protocol endianness and need to be converted to/from host endianness.

jhedberg replied:
Actually, looking at the use of these structs in the code it looks like you're missing the endianness conversions.

ekohandel replied:
These structures are populated by the `oacp_command_decode` function which uses the `net_buf_simple_pull_le` family of functions to do the endianness conversion. The reason they are being annotated with `__packed` attribute is that the `oacp_command_len_verify` function validates the procedure's request length against the size of these structures.

An alternate approach would be to remove the `__packed` attribute and provide definitions for the size of each procedure. I presume using `__packed` could have a performance impact on some architectures, would you rather see the definition approach?

jhedberg replied:
Yes, please only use __packed for structs which represent protocol-format data. Since we don't have endianness-annotated integer types like Linux this is the only way we can maintain at least a little bit of clarity and predictability for what endianness certain variables have.

joerchan commented on include/bluetooth/services/ots.h:
```diff
@@ -548,6 +548,31 @@ struct bt_ots_cb {
 			     uint64_t id, uint8_t **data, uint32_t len,
 			     uint32_t offset);
+
+	/** @brief Object write callback
```
It is unclear from this function if the application has to verify the offset and length attributes, (and what the application should verify if any) from this function.
By looking at the code it seems that the stack verifies length and offset.

I think it would make sense to document this.

ekohandel replied:
I have updated the function documentation to reflect the fact that the OTS module validates that the offset and length are within the allocated object size.

joerchan reviewed (commented):
Looks good. Just some suggestions in the comments.

jhedberg commented on include/bluetooth/services/ots.h:
```diff
@@ -548,6 +548,33 @@ struct bt_ots_cb {
+	 *                       moment. It should not be returned.
+	 */
+	int (*obj_write)(struct bt_ots *ots, struct bt_conn *conn, uint64_t id,
+			 uint8_t *data, uint32_t len, uint32_t offset,
```
I suppose `const uint8_t *data` would be more appropriate, since the callback is expected to treat the buffer as read-only?

jhedberg replied:
Btw, normally whenever we have "read" and "write" functionality we've tried to follow the POSIX read() and write() API parameter types where it makes sense, e.g. `ssize_t` return and `void *` instead of `uint8_t *` for the data. I suppose there were good reasons to deviate from this for the `obj_read()` and `obj_write()` callbacks?

ekohandel replied:
I was not aware of this desire but it makes sense. I don't think there is any reason for these APIs to deviate from the POSIX parameter types. I will change them to meet the suggested types.

ekohandel replied:
Changing the `obj_read()` function signature had a larger impact so I separated that into a separate PR to keep this PR focused on object write functionality: #36653.

I also opted to change the length and offset type parameters in these APIs to also use `size_t`.

jhedberg commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -134,11 +134,30 @@ static uint32_t ots_obj_read(struct bt_ots *ots, struct bt_conn *conn,
 }
 
+static int ots_obj_write(struct bt_ots *ots, struct bt_conn *conn,
+			     uint64_t id, uint8_t *data, uint32_t len,
```
Same thing here with the `const`

jhedberg reviewed (changes requested):
See my feedback that I provided as separate comments. The most sever issue seems to be missing endianness conversions, i.e. this code would likely be broken on a big-endian system (assuming you've been testing this on a little-endian architecture).

jhedberg commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -134,11 +134,30 @@ static uint32_t ots_obj_read(struct bt_ots *ots, struct bt_conn *conn,
+
+	memcpy(&objects[obj_index][offset], data, len);
+
+	return 0;
```
Shouldn't this be `return len` now that you switched to using ssize_t?

jhedberg replied:
To be clear: if you think "success or error" makes more sense as return value than "number of bytes consumed or error" then I'm not demanding you use ssize_t. That why I said "whenever it makes sense" in my original reference to POSIX APIs.

Thalley replied:
At this point I think it makes just as much sense to align this with the other Zephyr read/write APIs which does follow the POSIX APIs. To me it makes a whole lot of sense to have these APIs follow at least the L2CAP APIs since they are extending those.

ekohandel replied:
To be clear the current L2CAP receive API looks like below and doesn't have a `ssize_t` return type. It uses the "success or error" paradigm:
```
	/** @brief Channel recv callback
	 *
	 *  @param chan The channel receiving data.
	 *  @param buf Buffer containing incoming data.
	 *
	 *  @return 0 in case of success or negative value in case of error.
	 *  @return -EINPROGRESS in case where user has to confirm once the data
	 *                       has been processed by calling
	 *                       @ref bt_l2cap_chan_recv_complete passing back
	 *                       the buffer received with its original user_data
	 *                       which contains the number of segments/credits
	 *                       used by the packet.
	 */
	int (*recv)(struct bt_l2cap_chan *chan, struct net_buf *buf);
```
However, looks like the return handling for this would work even if we returned written length all the way back to the L2CAP layer:
```
	/* Receiving complete SDU, notify channel and reset SDU buf */
	err = chan->chan.ops->recv(&chan->chan, buf);
	if (err < 0) {
		if (err != -EINPROGRESS) {
			BT_ERR("err %d", err);
			bt_l2cap_chan_disconnect(&chan->chan);
			net_buf_unref(buf);
		}
		return;
	}
```
Although I am not sure what value this adds unless we check the return value against the actual incoming buffer length and terminate the connection on a mismatch? Or is there a more appropriate action to take?

From what I can tell, it looks like it makes sense to just keep this API as a success or error. If the application cannot handle the entire incoming buffer then it can reply with an error to terminate the channel.

Having said all of this, does it make sense to use `ssize_t` as the return type of this API? I am actually leaning towards going back to `int` to keep the API in line with the L2CAP APIs and be more clear that this is a "success or error" type API and not a "length or error" type API

Thalley replied:
I think it makes sense to look at the OTS read/write as a sort of extension of L2CAP, and thus I think it makes sense to keep in line with the L2CAP API. 

That being said, since its a GATT service, using the same API as GATT (which follows the POSIX standard) also makes sense. 

Preferably L2CAP had used POSIX and this would be easy to conclude on :) 

I think I'm slightly leaning towards @jhedberg first suggestion and use ssize_t for read and write, and have the functions return < 0 for error or the length written/read. 

@kapi-no You were the original implementer of OTS, do you have any input?

ekohandel replied:
Aligned the types to return `ssize_t`. < 0 is interpreted as an error and in the case of write if the returned length is not the same as the requested write length an `-EIO` error is propagated to the L2CAP layer. Alternatively I could have called back the write with the rest of the write buffer in the case that all of length was not written but it didn't seem to add much value as it would be called back immediately and likely whatever the reason was for being unable to do the write would still be present.

jhedberg commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -134,11 +134,30 @@ static uint32_t ots_obj_read(struct bt_ots *ots, struct bt_conn *conn,
 }
 
+static ssize_t ots_obj_write(struct bt_ots *ots, struct bt_conn *conn,
+			     uint64_t id, const void *data, uint32_t len,
+			     uint32_t offset, uint32_t rem)
```
Going with the POSIX mimicry approach would it make sense to consider `size_t len` and `off_t offset`?

jhedberg replied:
To be clear: this may or may not make sense. Using the POSIX types you abstract away from protocol details (e.g. if the protocol has 32-bit integers). So you should do this only if you think it makes sense to abstract away from protocol details.

ekohandel replied:
I think since the maximum size of the object is also a `uint32_t` and that is part of the protocol, it makes sense to keep these types as `uint32_t` to keep things aligned.

ekohandel replied:
I have changed the types to `size_t` and `off_t`.

carlescufi commented:
@ekohandel could you please reply to the type discussion so we can make progress?

ekohandel committed:
bluetooth: ots: Add object write procedure

Add the ability to perform an OTS Write Procedure with only Patch
support configuration.

Signed-off-by: Abe Kohandel <abe.kohandel@gmail.com>
